lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -569,6 +569,60 @@ resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE, DeclContext *DC) {
 
   // FIXME: Need to refactor the way we build an AST node from a lookup result!
 
+  SmallVector<ValueDecl*, 4> ResultValues;
+  ValueDecl *localDeclAfterUse = nullptr;
+  auto isValid = [&](ValueDecl *D) {
+    // FIXME: The source-location checks won't make sense once
+    // EnableASTScopeLookup is the default.
+    if (Loc.isValid() && D->getLoc().isValid() &&
+        D->getDeclContext()->isLocalContext() &&
+        D->getDeclContext() == DC &&
+        Context.SourceMgr.isBeforeInBuffer(Loc, D->getLoc())) {
```

Comment:
This check is necessary, but not sufficient for all cases. For example, a variable in a `guard` cannot be used within the `else` of that guard:

```
guard let x = someOptional else { /* x is not visible here */ }
```

However, I think that's okay: this pull request is about improving on the current situation, and as the FIXME notes, `ASTScope`-based name lookup is the full answer. More importantly, it's not *wrong* to do what this new code is doing, i.e., it won't reject valid code.